File: plugins/DEV-ESSENTIALS/hooks/test-file-reminder.js
#!/usr/bin/env node
/**
 * PostToolUse Hook: Test File Reminder
 *
 * When a new source file is created, checks if a corresponding
 * test file exists and reminds to write tests.
 *
 * Non-blocking (exit code 0).
 */

const fs = require('fs');
const path = require('path');
const MAX_STDIN = 1024 * 1024;

const SOURCE_PATTERNS = [
  { ext: /\.(ts|js)$/, testSuffix: ['.test.$1', '.spec.$1'] },
  { ext: /\.py$/, testSuffix: ['test_', '_test.py'] },
  { ext: /\.go$/, testSuffix: ['_test.go'] },
];

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('test-file-reminder')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const filePath = input.tool_input?.file_path || '';

    if (/\.(test|spec|_test)\./i.test(filePath) || /test_/i.test(path.basename(filePath))) {
      process.stdout.write(data);
      return;
    }

    if (!/\/(src|internal|lib|app|pkg|cmd)\//i.test(filePath) && !/^(src|internal|lib|app|pkg|cmd)\//i.test(filePath)) {
      process.stdout.write(data);
      return;
    }

    let isSource = false;
    let testPaths = [];

    for (const { ext, testSuffix } of SOURCE_PATTERNS) {
      if (ext.test(filePath)) {
        isSource = true;
        const dir = path.dirname(filePath);
        const base = path.basename(filePath);

        for (const suffix of testSuffix) {
          if (suffix.startsWith('test_')) {
            testPaths.push(path.join(dir, `test_${base}`));
          } else if (suffix.includes('$1')) {
            const replaced = base.replace(ext, suffix.replace('$1', base.match(ext)[0].substring(1)));
            testPaths.push(path.join(dir, replaced));
          } else {
            testPaths.push(path.join(dir, base.replace(ext, suffix)));
          }
        }
        break;
      }
    }

    if (!isSource) {
      process.stdout.write(data);
      return;
    }

    const hasTest = testPaths.some(tp => {
      try { return fs.existsSync(tp); } catch { return false; }
    });

    if (!hasTest) {
      console.error(`[Hook] REMINDER: No test file found for ${filePath}`);
      console.error(`[Hook] Expected test at: ${testPaths[0]}`);
      console.error('[Hook] Consider writing tests first (TDD: RED -> GREEN -> REFACTOR).');
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] test-file-reminder error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/DEV-ESSENTIALS/hooks/test-file-reminder.py
#!/usr/bin/env python3
"""
PostToolUse Hook: Test File Reminder

When a new source file is created, checks if a corresponding
test file exists and reminds to write tests.

Non-blocking (exit code 0).
"""

import json
import os
import re
import sys

MAX_STDIN = 1024 * 1024

SOURCE_PATTERNS = [
    (re.compile(r"\.(ts|js)$"), lambda base, ext: [
        re.sub(r"\.(ts|js)$", lambda m: f".test.{m.group(1)}", base),
        re.sub(r"\.(ts|js)$", lambda m: f".spec.{m.group(1)}", base),
    ]),
    (re.compile(r"\.py$"), lambda base, ext: [
        f"test_{base}",
        re.sub(r"\.py$", "_test.py", base),
    ]),
    (re.compile(r"\.go$"), lambda base, ext: [
        re.sub(r"\.go$", "_test.go", base),
    ]),
]

SOURCE_DIRS = re.compile(r"(?:^|/)(src|internal|lib|app|pkg|cmd)/", re.IGNORECASE)


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "test-file-reminder" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        file_path = inp.get("tool_input", {}).get("file_path", "")
        basename = os.path.basename(file_path)

        if re.search(r"\.(test|spec|_test)\.", file_path, re.IGNORECASE) or re.search(r"^test_", basename, re.IGNORECASE):
            sys.stdout.write(data)
            return

        if not SOURCE_DIRS.search(file_path):
            sys.stdout.write(data)
            return

        test_paths = []
        is_source = False
        dir_name = os.path.dirname(file_path)

        for ext_re, make_test_names in SOURCE_PATTERNS:
            if ext_re.search(file_path):
                is_source = True
                for test_name in make_test_names(basename, ext_re):
                    test_paths.append(os.path.join(dir_name, test_name))
                break

        if not is_source:
            sys.stdout.write(data)
            return

        has_test = any(os.path.exists(tp) for tp in test_paths)

        if not has_test:
            print(f"[Hook] REMINDER: No test file found for {file_path}", file=sys.stderr)
            print(f"[Hook] Expected test at: {test_paths[0]}", file=sys.stderr)
            print("[Hook] Consider writing tests first (TDD: RED -> GREEN -> REFACTOR).", file=sys.stderr)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] test-file-reminder error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
